src/hashset.c

```diff
@@ -0,0 +1,1681 @@
+    }
+    s->bucketExp[1] = exp;
+    s->tables[1] = new_table;
+    s->used[1] = 0;
+    s->rehashIdx = 0;
+    if (s->type->rehashingStarted) s->type->rehashingStarted(s);
+
+    /* If the old table was empty, the rehashing is completed immediately. */
+    if (s->tables[0] == NULL || s->used[0] == 0) {
+        rehashingCompleted(s);
+    } else if (s->type->instant_rehashing) {
+        while (hashsetIsRehashing(s)) {
+            rehashStep(s);
+        }
+    }
+    return 1;
+}
+
+/* Probing is slow when there are too many tombstones. Resize to the same size
+ * to trigger rehashing and cleaning up tombstones. */
+static int cleanUpTombstonesIfNeeded(hashset *s) {
+    if (hashsetIsRehashing(s) || resize_policy == HASHSET_RESIZE_FORBID) {
+        return 0;
+    }
+    if (s->everfulls[0] * 100 >= numBuckets(s->bucketExp[0]) * MAX_FILL_PERCENT_SOFT) {
+        return resize(s, s->used[0], NULL);
+    }
+    return 0;
+}
+
+/* Returns 1 if the table is expanded, 0 if not expanded. If 0 is returned and
+ * 'malloc_failed' is proveded, it is set to 1 if malloc failed and 0
+ * otherwise. */
+static int expand(hashset *s, size_t size, int *malloc_failed) {
+    if (size < hashsetSize(s)) {
+        return 0;
+    }
+    return resize(s, size, malloc_failed);
+}
+
+/* Finds an element matching the key. If a match is found, returns a pointer to
+ * the bucket containing the matching element and points 'pos_in_bucket' to the
+ * index within the bucket. Returns NULL if no matching element was found.
+ *
+ * If 'table_index' is provided, it is set to the index of the table (0 or 1)
+ * the returned bucket belongs to. */
+static bucket *findBucket(hashset *s, uint64_t hash, const void *key, int *pos_in_bucket, int *table_index) {
+    if (hashsetSize(s) == 0) return 0;
+    uint8_t h2 = highBits(hash);
+    int table;
+
+    /* Do some incremental rehashing. */
+    rehashStepOnReadIfNeeded(s);
+
+    /* Check rehashing destination table first, since it is newer and typically
+     * has less 'everfull' flagged buckets. Therefore it needs less probing for
+     * lookup. */
+    for (table = 1; table >= 0; table--) {
+        if (s->used[table] == 0) continue;
+        size_t mask = expToMask(s->bucketExp[table]);
+        size_t bucket_idx = hash & mask;
+        size_t start_bucket_idx = bucket_idx;
+        while (1) {
+            bucket *b = &s->tables[table][bucket_idx];
+            /* Find candidate elements with presence flag set and matching h2 hash. */
+            for (int pos = 0; pos < ELEMENTS_PER_BUCKET; pos++) {
+                if ((b->presence & (1 << pos)) && b->hashes[pos] == h2) {
+                    /* It's a candidate. */
+                    void *element = b->elements[pos];
+                    const void *elem_key = elementGetKey(s, element);
+                    if (compareKeys(s, key, elem_key) == 0) {
+                        /* It's a match. */
+                        if (pos_in_bucket) *pos_in_bucket = pos;
```

Comment:
```suggestion
                        *pos_in_bucket = pos;
```
Maybe add an `assert(pos_in_bucket)` since it seems like we always require this. This part of the code seems worthwhile to try to optimize as much as possible.